mixer/pkg/runtime2/handler/factory.go

```diff
@@ -0,0 +1,209 @@
+
+	"istio.io/api/mixer/v1/config/descriptor"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/expr"
+	"istio.io/istio/mixer/pkg/il/evaluator"
+	"istio.io/istio/mixer/pkg/runtime2/config"
+	"istio.io/istio/mixer/pkg/template"
+	"istio.io/istio/pkg/log"
+)
+
+// Map of instance name to inferred type (proto.Message)
+type inferredTypesMap map[string]proto.Message
+
+type factory struct {
+	snapshot *config.Snapshot
+
+	checker expr.TypeChecker
+
+	// protects cache
+	inferredTypesLock sync.RWMutex
+
+	// Map of instance name to inferred type (proto.Message)
+	inferredTypesCache map[string]proto.Message
+}
+
+func newFactory(snapshot *config.Snapshot) *factory {
+	return &factory{
+		snapshot: snapshot,
+		checker:  evaluator.NewTypeChecker(),
+
+		inferredTypesCache: make(map[string]proto.Message),
+	}
+}
+
+// build instantiates a handler object using the passed in handler and instances configuration.
+func (f *factory) build(
+	handler *config.Handler,
+	instances []*config.Instance,
+	env adapter.Env) (adapter.Handler, error) {
+
+	inferredTypesByTemplates, err := f.inferTypes(instances)
+	if err != nil {
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	// HandlerBuilder should always be present for a valid configuration (reference integrity should already be checked).
+	info := handler.Adapter
+
+	builder := info.NewBuilder()
+	if builder == nil {
+		msg := fmt.Sprintf("nil HandlerBuilder instantiated for adapter '%s' in handlerConfig config '%s'", handler.Adapter.Name, handler.Name)
+		log.Warn(msg)
+		return nil, errors.New(msg)
+	}
+
+	// validate if the builder supports all the necessary interfaces
+	for _, tmplName := range info.SupportedTemplates {
+		// TODO: !!! Is this true? Neither information is coming from config. It is not validated.
+		// ti should be there for a valid configuration.
+		ti, _ := f.snapshot.Templates[tmplName]
+
+		if supports := ti.BuilderSupportsTemplate(builder); !supports {
+			// adapter's builder is bad since it does not support the necessary interface
+			msg := fmt.Sprintf("adapter is invalid because it does not implement interface '%s'. "+
+				"Therefore, it cannot support template '%s'", ti.BldrInterfaceName, tmplName)
+			log.Error(msg)
+			return nil, fmt.Errorf(msg)
+		}
+	}
+
+	instantiatedAdapter, err := f.buildHandler(builder, inferredTypesByTemplates, handler.Params, env)
+	if err != nil {
+		msg := fmt.Sprintf("cannot configure adapter '%s' in handlerConfig config '%s': %v", handler.Adapter.Name, handler.Name, err)
+		log.Warn(msg)
+		return nil, errors.New(msg)
+	}
+	// validate if the handlerConfig supports all the necessary interfaces
+	for _, tmplName := range info.SupportedTemplates {
+		// ti should be there for a valid configuration.
+		ti, _ := f.snapshot.Templates[tmplName]
+
+		// TODO: !!! The adapter is instantiated. Should we simply drop it on the floor?
+		if supports := ti.HandlerSupportsTemplate(instantiatedAdapter); !supports {
```

Comment:
This particular check we cannto do during  adapter ingestion during getInfo calls. So, IMO we can cleanup this adapter completely from our known list of adapters, log and move all. Because it claimed it supported Foo template but didn't implement the right methods... (we can go fancy and mark the adapter-template combo to be bad but I think just getting rid of the complete adapter from our registry is not that bad. Makes our code simple)